# 🛠️ Project Setup

In this first step, we'll prepare the workspace for the **OKCupid Date-A-Scientist** project.

## 📦 Download the Project Materials

1. Download the ZIP file for the **Date-A-Scientist** project from the following link:  
   👉 [Download Starter Files](https://content.codecademy.com/PRO/paths/data-science/OKCupid-Date-A-Scientist-Starter.zip?_gl=1*1gdkc3o*_gcl_au*MTg1MTk2MTQ1LjE3NjE4MjEwMzk.*_ga*MTc5MTQ4OTgwMy4xNzMwMzc2Mzgy*_ga_3LRZM6TM9L*czE3NjM0MjgyMzYkbzEzNSRnMSR0MTc2MzQyOTQyNyRqNjAkbDAkaDA.)
2. Unzip the folder. You should see two items:
   - `date-a-scientist.ipynb` (a blank Jupyter Notebook)
   - `profiles.csv` (the dataset containing user profiles)

## 🚀 Launch Jupyter Notebook

1. Open your terminal or command line interface.
2. Type the following command to start Jupyter Notebook:

   ```bash
   jupyter notebook
   ```

3. A browser tab will open automatically.
4. Click on `date-a-scientist.ipynb` to open the notebook.
5. Build your project inside this file.

## 🧠 How to Use Jupyter Notebook

Jupyter Notebook is an interactive tool that lets you combine code, visualizations, and explanatory text in one document. It's perfect for data science projects because it supports exploration, analysis, and storytelling.

If you need help setting up or using Jupyter Notebook, check out these resources:

- [Command Line Interface Setup](#)
- [Introducing Jupyter Notebook](#)
- [Setting up Jupyter Notebook](#)
- [Getting Started with Jupyter](#)
- [Getting More out of Jupyter Notebook](#)

---


```markdown
# 🔐 Git Repository Setup with SSH and LFS

To manage version control and large files efficiently, this project uses **Git**, **SSH authentication**, and **Git LFS**.

## 📁 Local Project Path

```bash
cd code/DataScientistML_Codecademy/OKCupid-Date-A-Scientist-Starter
```

## 🧱 Initialize and Push to Remote Repository

```bash
git init
git lfs install
git add .
git commit -m "Initial commit for OKCupid Date-A-Scientist project"
git remote add origin git@personal-github:gabrielarcangelbol/OKCupid-Date-A-Scientist-Starter.git
git branch -M main
git push -u origin main
```

## ⚙️ Configure `.gitattributes` for Large Files

To track CSV files with Git LFS, the following line was added to `.gitattributes`:

```
*.csv filter=lfs diff=lfs merge=lfs -text
```

Then committed:

```bash
git add .gitattributes
git commit -m "Configure Git LFS for CSV files"
```

## 🔐 SSH Key Setup for Personal GitHub Account

To ensure Git uses your **personal SSH key**, follow these steps:

1. Start the SSH agent:

   ```bash
   eval "$(ssh-agent -s)"
   ```

2. Add your personal SSH key:

   ```bash
   ssh-add ~/.ssh/id_rsa_personal
   ```

## 🔍 Verify Remote Configuration

To confirm the correct remote is set:

```bash
cd code/DataScientistML_Codecademy/OKCupid-Date-A-Scientist-Starter
git remote -v
```

You should see:

```
origin  git@personal-github:gabrielarcangelbol/OKCupid-Date-A-Scientist-Starter.git (fetch)
origin  git@personal-github:gabrielarcangelbol/OKCupid-Date-A-Scientist-Starter.git (push)
```

---

✅ With Git, SSH, and LFS configured, your project is ready for version-controlled development and collaboration.
```

# 🧭 Project Scoping

Properly scoping your project creates structure and helps you think through your entire workflow before diving into the analysis. This section outlines the key components of the project scope, inspired by the [University of Chicago’s Data Science Project Scoping Guide](https://datasciencepublicpolicy.org/our-work/tools-guides/data-science-project-scoping-guide/).

---

## 🎯 1. Project Goals

- **Primary Objective**: Explore the OKCupid dataset to uncover patterns in dating preferences and behaviors using machine learning and NLP techniques.
- **Secondary Goals**:
  - Identify which user attributes are most predictive of compatibility.
  - Apply clustering to discover latent user segments.
  - Build a supervised model to predict user traits or preferences based on profile text.

---

## 📊 2. Data Overview

- **Source**: `profiles.csv` from OKCupid (provided in starter files)
- **Structure**: Each row represents a user profile with multiple features including:
  - Demographics (age, sex, orientation, location)
  - Lifestyle (diet, smoking, drinking, drugs)
  - Essay responses (10 free-text fields)
- **Challenges**:
  - Missing values
  - Unstructured text (essays)
  - Potential class imbalance

---

## 🧪 3. Analytical Approach

### Exploratory Data Analysis (EDA)
- Distribution of key demographics
- Missing data patterns
- Word frequency and sentiment in essay fields

### Feature Engineering
- Text vectorization (TF-IDF, embeddings)
- Aggregated lifestyle indicators
- Derived compatibility scores

### Modeling
- **Unsupervised**: Clustering (e.g., KMeans, DBSCAN) to identify user segments
- **Supervised**: Classification (e.g., logistic regression, random forest) to predict:
  - Smoking habits
  - Orientation
  - Personality traits (inferred from essays)

---

## 🧱 4. Constraints & Assumptions

- **Assumptions**:
  - Users are honest in their profiles
  - Essay content reflects personality and preferences
- **Constraints**:
  - No access to actual match outcomes
  - Limited metadata on user interactions

---

## 🔄 5. Risks & Adjustments

- NLP models may underperform due to short or noisy text
- Clustering may not yield interpretable segments
- Some hypotheses may not be supported by the data

---

## 📌 Next Steps

- Clean and preprocess the dataset
- Perform EDA and visualize key trends
- Define target variables for modeling
- Iterate on feature engineering and model evaluation

---

📚 For more guidance, refer to the full [Data Science Project Scoping Guide](https://datasciencepublicpolicy.org/our-work/tools-guides/data-science-project-scoping-guide/) and [Scoping Worksheet (PDF)](https://worldbank.github.io/Data-in-Action/downloads/UChicago_DSSG_Project_Scoping.pdf).

Sources: [Scoping Guide](https://datasciencepublicpolicy.org/our-work/tools-guides/data-science-project-scoping-guide/), [Scoping Worksheet PDF](https://worldbank.github.io/Data-in-Action/downloads/UChicago_DSSG_Project_Scoping.pdf)


# 🤖 Select ML-Solvable Problem

In this section, we define a machine learning problem that is feasible given the dataset and project scope. The goal is to identify a task that:

- Can be solved using available data
- Is achievable within a reasonable timeframe
- Has a clear input-output structure suitable for ML
- Aligns with the broader goals of the project

---

## 🧩 Problem Statement

Many users on OKCupid consider **Zodiac signs** important when evaluating compatibility. However, not all users provide their zodiac sign in their profile. This creates a gap in the matching process.

**Problem**: Can we predict a user's zodiac sign based on other profile attributes such as lifestyle habits (drinking, smoking, drug use) and free-text essay responses?

---

## 🧠 ML Framing

This is a **multi-class classification** problem with 12 possible zodiac labels.

### **Input Features**:
- Categorical: `drinks`, `smokes`, `drugs`, `job`, `education`
- Textual: `essay0` to `essay9` (to be vectorized using NLP techniques)

### **Target Variable**:
- `sign` (Zodiac sign)

---

## 🔍 Why This Problem?

- The `sign` column has missing values, making it a good candidate for imputation via ML.
- Lifestyle and personality traits (expressed in essays) may correlate with astrological archetypes.
- Predicting zodiac signs could enhance match recommendations for users who omit this field.

---

## 🧪 ML Techniques to Be Used

- **Text preprocessing**: cleaning, tokenization, TF-IDF or embeddings
- **Feature encoding**: one-hot or ordinal encoding for categorical variables
- **Modeling**: 
  - Baseline: Logistic Regression or Naive Bayes
  - Advanced: Random Forest, XGBoost, or fine-tuned transformer (if time permits)
- **Evaluation**: Accuracy, F1-score, confusion matrix

---

## 📌 Next Steps

- Analyze class distribution of `sign`
- Explore correlations between zodiac and lifestyle features
- Preprocess essays and engineer features
- Train and evaluate classification models

---

📚 Reference: [Google’s Guide to Identifying ML-Suitable Problems](https://developers.google.com/machine-learning/problem-framing)

# 📂 Load and Check Data

Before applying any machine learning techniques, we need to load the dataset and verify that it contains a suitable **label or response variable** for supervised learning.

The dataset is stored in `profiles.csv` and includes:

### 🧾 Multiple-choice columns:
- `body_type`, `diet`, `drinks`, `drugs`, `education`, `ethnicity`, `height`, `income`, `job`, `offspring`, `orientation`, `pets`, `religion`, `sex`, `sign`, `smokes`, `speaks`, `status`

### 📝 Essay fields:
- `essay0` to `essay9`, covering topics like self-summary, lifestyle, preferences, and personality

---

## 🎯 Target Variable Check

Since our selected ML task is to **predict the user's zodiac sign**, we will check the `sign` column for:

- Presence of values
- Class distribution
- Missing data

If the `sign` column is too sparse or unreliable, we may need to revisit our problem definition.

---

## 📦 Load Dataset
```python
import pandas as pd

# Load the dataset
df = pd.read_csv("profiles.csv")

# Display basic info
print("Shape of dataset:", df.shape)
df.info()

# Preview the first few rows
df.head()
```

---

## ✅ Next Steps

- Check for missing values in the `sign` column
- Explore distribution of zodiac signs
- Assess completeness of other relevant features (`drinks`, `smokes`, `drugs`, `essay0`–`essay9`)
- Decide whether to proceed with this target or redefine the ML problem

In [3]:
# 📦 Load and Inspect OKCupid Dataset

import pandas as pd

# Load the dataset
df = pd.read_csv("profiles.csv")

# Basic shape and structure
print("✅ Dataset loaded successfully.")
print("🔢 Number of rows and columns:", df.shape)

# Overview of column types and missing values
print("\n📋 Dataset Info:")
df.info()

# Preview the first few rows
print("\n🔍 First 5 rows:")
df.head()

✅ Dataset loaded successfully.
🔢 Number of rows and columns: (59946, 31)

📋 Dataset Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59946 entries, 0 to 59945
Data columns (total 31 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   age          59946 non-null  int64  
 1   body_type    54650 non-null  object 
 2   diet         35551 non-null  object 
 3   drinks       56961 non-null  object 
 4   drugs        45866 non-null  object 
 5   education    53318 non-null  object 
 6   essay0       54458 non-null  object 
 7   essay1       52374 non-null  object 
 8   essay2       50308 non-null  object 
 9   essay3       48470 non-null  object 
 10  essay4       49409 non-null  object 
 11  essay5       49096 non-null  object 
 12  essay6       46175 non-null  object 
 13  essay7       47495 non-null  object 
 14  essay8       40721 non-null  object 
 15  essay9       47343 non-null  object 
 16  ethnicity    54266 non-null  object 
 17

age       body_type               diet    drinks      drugs  \
0   22  a little extra  strictly anything  socially      never   
1   35         average       mostly other     often  sometimes   
2   38            thin           anything  socially        NaN   
3   23            thin         vegetarian  socially        NaN   
4   29        athletic                NaN  socially      never   

                           education  \
0      working on college/university   
1              working on space camp   
2     graduated from masters program   
3      working on college/university   
4  graduated from college/university   

                                              essay0  \
0  about me:<br />\n<br />\ni would love to think...   
1  i am a chef: this is what that means.<br />\n1...   
2  i'm not ashamed of much, but writing public te...   
3          i work in a library and go to school. . .   
4  hey how's it going? currently vague on the pro...   

                                              essay1  \
0  currently working as an international agent fo...   
1  dedicating everyday to being an unbelievable b...   
2  i make nerdy software for musicians, artists, ...   
3          reading things written by old dead people   
4                         work work work work + play   

                                              essay2  \
0  making people laugh.<br />\nranting about a go...   
1  being silly. having ridiculous amonts of fun w...   
2  improvising in different contexts. alternating...   
3  playing synthesizers and organizing books acco...   
4  creating imagery to look at:<br />\nhttp://bag...   

                                              essay3  ...  \
0  the way i look. i am a six foot half asian, ha...  ...   
1                                                NaN  ...   
2  my large jaw and large glasses are the physica...  ...   
3                  socially awkward but i do my best  ...   
4            i smile a lot and my inquisitive nature  ...   

                          location  \
0  south san francisco, california   
1              oakland, california   
2        san francisco, california   
3             berkeley, california   
4        san francisco, california   

                                      offspring orientation  \
0  doesn&rsquo;t have kids, but might want them    straight   
1  doesn&rsquo;t have kids, but might want them    straight   
2                                           NaN    straight   
3                       doesn&rsquo;t want kids    straight   
4                                           NaN    straight   

                        pets                                  religion sex  \
0  likes dogs and likes cats     agnosticism and very serious about it   m   
1  likes dogs and likes cats  agnosticism but not too serious about it   m   
2                   has cats                                       NaN   m   
3                 likes cats                                       NaN   m   
4  likes dogs and likes cats                                       NaN   m   

                                 sign     smokes  \
0                              gemini  sometimes   
1                              cancer         no   
2  pisces but it doesn&rsquo;t matter         no   
3                              pisces         no   
4                            aquarius         no   

                                              speaks     status  
0                                            english     single  
1  english (fluently), spanish (poorly), french (...     single  
2                               english, french, c++  available  
3                           english, german (poorly)     single  
4                                            english     single  

[5 rows x 31 columns]

In [4]:
# 📊 Check Target Variable and Feature Completeness

# Check how many missing values are in the 'sign' column
missing_sign = df['sign'].isnull().sum()
total_rows = len(df)
print(f"❓ Missing 'sign' values: {missing_sign} out of {total_rows} ({missing_sign/total_rows:.2%})")

# View distribution of zodiac signs
print("\n🔮 Zodiac Sign Distribution:")
print(df['sign'].value_counts(dropna=False))

# Check missing values in lifestyle features
print("\n🚦 Missing values in lifestyle columns:")
print(df[['drinks', 'smokes', 'drugs']].isnull().sum())

# Check completeness of essay fields
print("\n📝 Missing values in essay fields:")
essay_cols = [f"essay{i}" for i in range(10)]
print(df[essay_cols].isnull().sum())


❓ Missing 'sign' values: 11056 out of 59946 (18.44%)

🔮 Zodiac Sign Distribution:
sign
NaN                                              11056
gemini and it&rsquo;s fun to think about          1782
scorpio and it&rsquo;s fun to think about         1772
leo and it&rsquo;s fun to think about             1692
libra and it&rsquo;s fun to think about           1649
taurus and it&rsquo;s fun to think about          1640
cancer and it&rsquo;s fun to think about          1597
pisces and it&rsquo;s fun to think about          1592
sagittarius and it&rsquo;s fun to think about     1583
virgo and it&rsquo;s fun to think about           1574
aries and it&rsquo;s fun to think about           1573
aquarius and it&rsquo;s fun to think about        1503
virgo but it doesn&rsquo;t matter                 1497
leo but it doesn&rsquo;t matter                   1457
cancer but it doesn&rsquo;t matter                1454
gemini but it doesn&rsquo;t matter                1453
taurus but it doesn&rsquo;t matte

# 🔍 Preliminary Data Assessment

After loading and inspecting the dataset, we identified several key insights and challenges that will guide our next steps.

## 🧠 Target Variable: `sign`

- ✅ Present in 48,890 out of 59,946 rows (~81.56%)
- ❌ Missing in ~18.44% of entries
- ✅ Contains 12 zodiac signs, but with many variations (e.g., "gemini and it’s fun to think about", "gemini but it doesn’t matter", "gemini")
- 🔧 **Action**: Normalize zodiac labels by extracting the base sign (e.g., "gemini") and discarding modifiers

## 🚦 Lifestyle Features

| Feature   | Missing Values | % Missing |
|-----------|----------------|-----------|
| `drinks`  | 2,985          | ~5.0%     |
| `smokes`  | 5,512          | ~9.2%     |
| `drugs`   | 14,080         | ~23.5%    |

- 🔧 **Action**: Consider imputing missing values or treating them as a separate category ("unknown")

## 📝 Essay Fields

- Missing values range from ~9% (`essay0`) to ~32% (`essay8`)
- Text fields are rich but sparse and noisy
- 🔧 **Action**:
  - Combine essays into a single text field for NLP preprocessing
  - Apply cleaning (lowercasing, punctuation removal, etc.)
  - Use TF-IDF or embeddings for feature extraction

## ⚠️ Other Observations

- `offspring`, `pets`, and `religion` have high missingness (>30%)
- `income` is present but may be skewed or zero-filled
- `height` and `age` are complete and usable

---

## ✅ Next Steps

1. **Normalize the `sign` column** to extract base zodiac labels
2. **Clean and combine essay fields** for NLP feature engineering
3. **Handle missing values** in lifestyle and categorical features
4. **Visualize distributions** to guide feature selection
5. **Prepare training data** for supervised classification

---

📌 These steps will help us build a robust pipeline for predicting zodiac signs based on lifestyle and personality traits.


# 🧹 Data Cleaning: Zodiac Signs and Essay Text

Before we begin exploratory analysis, we need to clean and standardize key features in the dataset. This includes:

1. Normalizing the `sign` column to extract the base zodiac label
2. Combining all essay fields into a single text column for NLP processing

These steps will ensure consistency and simplify downstream modeling.


### ✅ Step 1: Normalize the `sign` column

In [5]:
import re
import pandas as pd

# Define a function to extract the base zodiac sign from messy strings
def extract_zodiac(sign):
    if pd.isnull(sign):
        return None
    # Use regex to find the zodiac keyword in the string
    match = re.search(r'\b(aries|taurus|gemini|cancer|leo|virgo|libra|scorpio|sagittarius|capricorn|aquarius|pisces)\b', sign.lower())
    return match.group(0) if match else None

# Apply the function to the 'sign' column
df['zodiac_clean'] = df['sign'].apply(extract_zodiac)

# Display the cleaned distribution
print("🔮 Cleaned Zodiac Sign Distribution:")
print(df['zodiac_clean'].value_counts(dropna=False))

🔮 Cleaned Zodiac Sign Distribution:
zodiac_clean
None           11056
leo             4374
gemini          4310
libra           4207
cancer          4206
virgo           4141
taurus          4140
scorpio         4134
aries           3989
pisces          3946
sagittarius     3942
aquarius        3928
capricorn       3573
Name: count, dtype: int64


**Explanation:**
- Many entries in the `sign` column contain modifiers like “but it doesn’t matter” or “and it’s fun to think about”.
- This function extracts only the core zodiac sign using regular expressions.
- The result is stored in a new column `zodiac_clean`.
---


### ✅ Step 2: Combine all essay fields into one column

In [6]:
# List of essay column names
essay_cols = [f"essay{i}" for i in range(10)]

# Fill missing essay values with empty strings and concatenate them
df['essays_combined'] = df[essay_cols].fillna('').agg(' '.join, axis=1)

# Preview the combined essay text for the first profile
print("📝 Sample combined essay text:")
print(df['essays_combined'].iloc[0][:500])  # Show first 500 characters

📝 Sample combined essay text:
about me:<br />
<br />
i would love to think that i was some some kind of intellectual:
either the dumbest smart guy, or the smartest dumb guy. can't say i
can tell the difference. i love to talk about ideas and concepts. i
forge odd metaphors instead of reciting cliches. like the
simularities between a friend of mine's house and an underwater
salt mine. my favorite word is salt by the way (weird choice i
know). to me most things in life are better as metaphors. i seek to
make myself a little be


**Explanation:**
- Essay fields are spread across `essay0` to `essay9`, each representing a different prompt.
- We fill missing values with empty strings to avoid `NaN` issues.
- Then we concatenate all essays into a single column `essays_combined` for easier NLP processing.

---